---
title: "revising"
output: html_document
date: "2024-12-27"
---

rm(list = ls())
```{r libaries, include=FALSE}

options(scipen = 1000000000)
options(stringsAsFactors = F)
set.seed(1009)
library(readxl)
library(dplyr)
library(corrplot)
library(corrr)
library(RColorBrewer)
library(psych)
library(GGally)

library(CBCgrps)
library(tibble)
library(ggrepel)
library(ggplot2)
library(hrbrthemes)
library(ggpubr)

library(caret)
library(randomForest)
library(pROC)
library(e1071)
library(kknn)
library(adabag)
packageVersion("caret")
```


```{r correlation analysis, include=FALSE}
setwd("C:/Users/JiefeiNiu/OneDrive - Helmholtz Zentrum München/project/project/pro/onlypro-final/revising")
com_data <- read.csv("pheno_pro_2045.csv",check.names = FALSE)
data <- read.csv(file = "pheno_11_pro_809.csv", row.names = 1, check.names = FALSE)
data1 <- data.frame(row.names = com_data$zz_nr,
                    BMI = com_data$u3tbmi,
                    WHR = com_data$u3twhrat,
                    WC = com_data$u3ttumf,
                    BFP = com_data$u3tbfp_k,
                    age = com_data$u3talteru,
                    SBP = com_data$u3tsysmm,
                    DBP = com_data$u3tdiamm,
                    alcohol = com_data$u3talkkon,
                    FPG = com_data$u3tglukfast_n,
                    TG =  com_data$u3l_trin,
                    HDL = com_data$u3l_hdln,
                    LDL = com_data$u3l_ldln,
                    TCHO = com_data$u3l_choln) # CRP = com_data$u3h_crp
pro <- read.csv(file = "select_pro_agesex.csv", row.names = 1, check.names = FALSE)
data2 <- pro[, -c(1:3)]

cor.result <- corr.test(data2, data1, use = "pairwise.complete.obs", method = "spearman")
r <- cor.result$r
p <- cor.result$p
#write.csv(r, file = "pheno_pro_correlation_R.csv")
#write.csv(p, file = "pheno_pro_correlation_pvalue.csv")

head(data2)
#correlation in proteins and filter by 0.5
a <- cor(data2)


p[ p >= (0.05)] <- NA
rowSums(!is.na(p))
colSums(!is.na(p))
x <- data.frame(r[p < 0.05])
r <- cor.result$r
max(r)
min(r)
p <- cor.result$p

#col <- colorRampPalette(c("#BB4444", "#EE9988", "#FFFFFF", "#77AADD", "#4477AA"))(10) #"#EE9988", "#FFFFFF",
col <- colorRampPalette(c("blue", "white", "red"))(10)
#col <- colorRampPalette(c("darkred","white","midnightblue"))(10)
dev.new()
png("pheno_pro_cor_nocrp.png", width=27, height=21, units = "cm", res=600) 
p <- cor.result$p
corrplot(r, method = "square", p.mat = p, sig.level = 0.05,
         insig = "blank", tl.col = "black", col =col, tl.cex = 1.5)
dev.off()

cor.result1 <- corr.test(data1[, 1:4], use = "pairwise.complete.obs", method = "spearman")
r1 <- cor.result1$r
p1 <- cor.result1$p
#write.csv(r1, file = "obesity_correlation_R.csv")
#write.csv(p1, file = "obesity_correlation_pvalue.csv")
dev.new()
png("obesity_cor.png", width=27, height=21, units = "cm", res=600) 
corrplot(r1, method = "square", p.mat = p1, sig.level = 0.05,
         insig = "blank", tl.col = "black", col =col, tl.cex = 1.5)
dev.off()

```

```{r RF}
rm(list = ls())
data <- read.csv(file = "select_pro_agesex.csv", row.names = 1) #don't , check.names = FALSE here
data <- data[,-c(2,3)] #delete sex and age colums
# change "male"and "female"to numeric
#data$sex <- ifelse(data$sex == "male", 1, 0)
data$BMI <- as.factor(data$BMI)
set.seed(1009)
ind<-sample(2,nrow(data),replace = TRUE,prob = c(0.7,0.3))
traindata<-data[ind == 1,]
testdata<-data[ind == 2,]


set.seed(1009)
#default is 500 trees
randomForest_rf<-randomForest(BMI ~.,data = traindata, #BMI ~ . - age - sex
                              ntree = 500,
                              mtry = 5,
                              proximity = TRUE,
                              importance = TRUE)
#model validation
set.seed(1009)
irispred <-predict(randomForest_rf,newdata = testdata)
freq1 <- table(irispred,testdata$BMI)
#plot(margin(randomForest_rf,testdata$BMI))
#mixed matrix
caret::confusionMatrix(testdata$BMI, irispred) #AUC=0.7841

irispred1 <- predict(randomForest_rf,newdata = testdata, type = "pro")
rocc1 <- roc(testdata$BMI, as.matrix(irispred1[ , 1]),ci=T,auc=T, direction=">")
rocc1


#python code
#refrence the link: https://blog.csdn.net/sinat_26917383/article/details/115400327
#reference the paper: https://www.nature.com/articles/s41591-023-02396-3 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import shap
import time
import numba
import sklearn.ensemble
import os


os.chdir('C:/Users/JiefeiNiu/OneDrive - Helmholtz Zentrum München/project/project/pro/onlypro-final/revising') #set the working directory
dt = pd.read_csv('C:/Users/JiefeiNiu/OneDrive - Helmholtz Zentrum München/project/project/pro/onlypro-final/revising/select_pro_agesex.csv')
from sklearn.preprocessing import LabelEncoder
dt.BMI
dt
dt = dt.drop(dt.columns[[0, 2, 3]], axis=1) #delete id, sex, age
dt['BMI'] = dt['BMI'].astype('category') #change the type of BMI to category
dt.dtypes
dt.head()
np.random.seed(1009)

from sklearn.model_selection import train_test_split
X = dt.drop('BMI', axis=1)
X = pd.get_dummies(X, drop_first=True)
y = dt['BMI']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1009)

model = sklearn.ensemble.RandomForestClassifier(n_estimators=500, max_features=5, oob_score=True, random_state=1009)
model.fit(X_train, y_train)

from sklearn.metrics import accuracy_score, classification_report
y_pred = model.predict(X_test) 
accuracy = accuracy_score(y_test, y_pred) 
report = classification_report(y_test, y_pred) 
print(f'Accuracy: {accuracy}') 
print(report)

explainer = shap.TreeExplainer(model)
shap_values = explainer(X)
shap_values[0]
shap_values = shap_values[..., 1]

feature_importance = np.mean(np.abs(shap_values.values), axis=0)
df1 = pd.DataFrame(feature_importance, index=X.columns, columns=['Importance'])
df1 = df1.sort_values('Importance', ascending=False)
df1
df1.to_csv('C:/Users/JiefeiNiu/OneDrive - Helmholtz Zentrum München/project/project/pro/onlypro-final/revising/importance.csv')

shap.plots.beeswarm(shap_values, max_display=16, color_bar=True, s=16, plot_size='auto', color_bar_label='Feature value')

shap.plots.waterfall(shap_values[1]) # the firtst sample
shap.plots.bar(shap_values, max_display=16) #summarize the effects of all the features
shap.plots.heatmap(shap_values)
shap.plots.force(explainer.expected_value,shap_values.values,shap_values.data)
shap.plots.force(shap_values) 


# summarize the effects of all the features
shap.plots.beeswarm(shap_values, max_display=16, color_bar=True, s=16, plot_size='auto',color_bar_label='Feature value')
plt.close() 
shap.plots.beeswarm(shap_values, max_display=16, show=False)
plt.tick_params(axis='both', labelsize=12)
colorbar = plt.gca().collections[-1].colorbar  # Get the color bar
plt.savefig('C:/Users/JiefeiNiu/OneDrive - Helmholtz Zentrum München/project/project/pro/onlypro-final/revising/SHAP_beeswarm.png',dpi = 600, bbox_inches = 'tight', pad_inches=0.1)
plt.show() 
plt.close() 

shap.plots.beeswarm(shap_values, max_display=28, show=False)
plt.savefig('C:/Users/JiefeiNiu/OneDrive - Helmholtz Zentrum München/project/project/pro/onlypro-final/revising/SHAP_allbeeswarm.png',dpi = 600, bbox_inches = 'tight', pad_inches=0.1)
plt.show() 
plt.close() 

plt.close() 
shap.plots.bar(shap_values, max_display=16, show=False)
plt.tick_params(axis='both', labelsize=12)
plt.savefig('C:/Users/JiefeiNiu/OneDrive - Helmholtz Zentrum München/project/project/pro/onlypro-final/revising/SHAP_barplot.png',dpi = 600, bbox_inches = 'tight', pad_inches=0.1)
plt.show() 
plt.close()

plt.close() 
shap.plots.heatmap(shap_values, max_display=16, show=False)
plt.tick_params(axis='both', labelsize=12)
plt.savefig('C:/Users/JiefeiNiu/OneDrive - Helmholtz Zentrum München/project/project/pro/onlypro-final/revising/SHAP_heatmap.png',dpi = 600, bbox_inches = 'tight', pad_inches=0.1)
plt.show() 
plt.close()

```


```{r MR}
library(TwoSampleMR)
library(MRInstruments)
library(ggplot2)
library(data.table)
library(readr)
set.seed(1234)

rm(list = ls())
###################prepare proteins GWAS data#######################
folder_path <- "C:/Users/JiefeiNiu/Desktop/sun/"
#folder_path <- "/home/ame/jiefei.niu/tools/project/pQTL/" #put in server

# extract files' name
file_names <- list.files(folder_path)[1]
#file_names <- c(file_names[5])

result_list <- list()
for (file in file_names) {
  file_path <- paste0("C:/Users/JiefeiNiu/Desktop/sun/", file)
  data <- fread(file_path)
  #p_value < 5e-08 or logpvalue >8
  ex <- subset(data, p_value < 5e-08)
  # add files name as a column
  ex$FileName <- gsub(".tsv", "", file)
  # put results in list
  result_list[[file]] <- ex
}

#combine all res together
colnames(data)
head(data)
result <- rbindlist(result_list)
colnames(result)
#save it
#write.csv(result, file = "C:/project/pro/onlypro/result/mr/result.csv")
#write.csv(result, file = "C:/project/pro/onlypro/result/mr/result_LGALS3BP.csv")


###########################proteins lead to BMI###################################
#Ferkingstad, E. et al. Large-scale integration of the plasma proteome with genetics and disease. 
result <- read.csv("result.csv")
#check data
colnames(result)
table(result$genename)
head(result)
#include 9 PROTEINS: A2M, APCS, ADIPOQ, AFM, APOD, AZGP1, CRP, S100A9, SERPINA6
result <- subset(result, genename != "SERPINA6") #SERPINA6 cann't clumped
exp_dat <- format_data(
  result,
  type = "exposure",
  header = TRUE,
  snp_col = "rsids",
  beta_col = "Beta",
  se_col = "SE",
  pval_col = "Pval",
  effect_allele_col = "effectAllele",
  other_allele_col = "otherAllele",
  ncase_col = "N",
  id_col = "genename")

#APOF "prot-a-133"; APOM "prot-a-136"; CFH "prot-a-520"; A2M 'prot-a-1781';
#SERPINF1 "prot-a-2698"; 
#put the id one by one and get the results
library(ieugwasr)
exp_dat <- extract_instruments(outcomes = 'prot-a-133')
exp_dat <- extract_instruments(outcomes = 'prot-a-136')
exp_dat <- extract_instruments(outcomes = 'prot-a-520')
exp_dat <- extract_instruments(outcomes = 'prot-a-1781')
exp_dat <- extract_instruments(outcomes = 'prot-a-2698')

snp_exp_dat <- clump_data(exp_dat, clump_kb = 10000, clump_r2 = 0.001,
                          clump_p1 = 1, clump_p2 = 1, pop = "EUR")
# outcomes data
#BMI ebi-a-GCST90029007  n:532,396	SNP:11,973,091
#obesity ebi-a-GCST001475 n:13,848	SNP: 2,430,514
out_dat <- extract_outcome_data(
  snps = snp_exp_dat$SNP,
  outcomes = 'ebi-a-GCST001475'
)

#harmonise data
dat <- harmonise_data(
  exposure_dat = snp_exp_dat, 
  outcome_dat = out_dat)
#write.csv(dat, file = "C:/project/pro/onlypro-final/result/mr/result_snps.csv")
res <- mr(dat, method_list = c("mr_wald_ratio", "mr_ivw")) #"mr_egger_regression",
res
#Sensitivity analyses 
#Heterogeneity statistics
mr_heterogeneity(dat, method_list = c("mr_ivw", "mr_egger_regression"))
mr_heterogeneity(dat, method_list = c("mr_ivw"))
#可以挑出离群值
#run_mr_presso(dat, NbDistribution = 3000)
#Horizontal pleiotropy
mr_pleiotropy_test(dat)

#PLOT
res_single=mr_singlesnp(dat)
mr_forest_plot(res_single)

mr_funnel_plot(singlesnp_results = res_single)
mr_leaveoneout_plot(leaveoneout_results = mr_leaveoneout(dat))

#write.csv(snp_exp_dat, file = "C:/project/pro/onlypro/result/mr/snp_exp_dat.csv")
#write.csv(res, file = "C:/project/pro/onlypro/result/mr/res.csv")

#from LGALS3BP to BMI
result <- read.csv("result_LGALS3BP.csv")
exp_dat <- format_data(
  result,
  type = "exposure",
  header = TRUE,
  snp_col = "rs_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  ncase_col = "n")

snp_exp_dat <- clump_data(exp_dat, clump_kb = 10000, clump_r2 = 0.001,
                          clump_p1 = 1, clump_p2 = 1, pop = "EUR")
#write.csv(snp_exp_dat, file = "C:/project/pro/onlypro/result/mr/snp_exp_dat.csv")
# outcomes data
out_dat <- extract_outcome_data(
  snps = snp_exp_dat$SNP,
  outcomes = 'ebi-a-GCST90029007'
)

#harmonise data
dat <- harmonise_data(
  exposure_dat = snp_exp_dat, 
  outcome_dat = out_dat)
#write.csv(dat, file = "C:/project/pro/onlypro-final/result/mr/result_snps.csv")
res <- mr(dat, method_list = c("mr_wald_ratio", "mr_ivw"))
res
#Sensitivity analyses 
#Heterogeneity statistics
mr_heterogeneity(dat, method_list = c("mr_ivw", "mr_egger_regression"))
mr_heterogeneity(dat, method_list = c("mr_ivw"))
#可以挑出离群值
#run_mr_presso(dat, NbDistribution = 3000)
#Horizontal pleiotropy
mr_pleiotropy_test(dat)



###########################BMI lead to proteins###################################
#exposure
exp_dat1 <-extract_instruments(outcomes = 'ebi-a-GCST90029007')
snp_exp_dat1 <- clump_data(exp_dat1, clump_kb = 10000, clump_r2 = 0.001,
                           clump_p1 = 1, clump_p2 = 1, pop = "EUR")
#outcome
#include 9 PROTEINS: A2M, APCS, ADIPOQ, AFM, APOD, AZGP1, CRP, S100A9, SERPINA6
out_dat1 <- read_outcome_data(
  snps = snp_exp_dat1$SNP,
  filename = "C:/project/pro/onlypro-final/result/mr/result.csv",
  sep = ",",
  snp_col = "rsids",
  beta_col = "Beta",
  se_col = "SE",
  pval_col = "Pval",
  effect_allele_col = "effectAllele",
  other_allele_col = "otherAllele",
  ncase_col = "N",
  id_col = "genename"
)

#harmonise data
dat <- harmonise_data(
  exposure_dat = snp_exp_dat1, 
  outcome_dat = out_dat1)

res <- mr(dat, method_list = c("mr_wald_ratio", "mr_ivw"))
res
#Sensitivity analyses 
#Heterogeneity statistics
mr_heterogeneity(dat, method_list = c("mr_ivw"))
#可以挑出离群值
#run_mr_presso(dat, NbDistribution = 3000)
#Horizontal pleiotropy
mr_pleiotropy_test(dat)

#PLOT
res_single=mr_singlesnp(dat)
mr_forest_plot(res_single)

mr_funnel_plot(singlesnp_results = res_single)
mr_leaveoneout_plot(leaveoneout_results = mr_leaveoneout(dat))


#directly find GWAS data from GWAS catalog
#APOF "prot-a-133"; APOM "prot-a-136"; CFH "prot-a-520"; A2M 'prot-a-1781'
out_dat1 <- extract_outcome_data(snps = snp_exp_dat1$SNP, outcomes = 'prot-a-133')
out_dat1 <- extract_outcome_data(snps = snp_exp_dat1$SNP, outcomes = 'prot-a-136')
out_dat1 <- extract_outcome_data(snps = snp_exp_dat1$SNP, outcomes = 'prot-a-520')
out_dat1 <- extract_outcome_data(snps = snp_exp_dat1$SNP, outcomes = 'prot-a-1781')
#harmonise data
dat <- harmonise_data(
  exposure_dat = snp_exp_dat1, 
  outcome_dat = out_dat1)

res <- mr(dat, method_list = c("mr_wald_ratio", "mr_ivw"))
res
#Sensitivity analyses 
#Heterogeneity statistics
mr_heterogeneity(dat, method_list = c("mr_ivw"))
#可以挑出离群值
#run_mr_presso(dat, NbDistribution = 3000)
#Horizontal pleiotropy
mr_pleiotropy_test(dat)

#PLOT
res_single=mr_singlesnp(dat)
mr_forest_plot(res_single)

mr_funnel_plot(singlesnp_results = res_single)
mr_leaveoneout_plot(leaveoneout_results = mr_leaveoneout(dat))

```


```{r pressure, echo=FALSE}
plot(pressure)
```

